ExoPlayer.Core/Additions/DownloadService.cs

```diff
@@ -0,0 +1,577 @@
+
+        public static string action_start_downloads
+        {
+            get { return ACTION_START_DOWNLOADS; }
+        }
+
+        public static string action_stop_downloads
+        {
+            get { return ACTION_STOP_DOWNLOADS; }
+        }
+
+
+
+        /** Starts a download service without adding a new {@link DownloadAction}. */
+        public const string ACTION_INIT = "com.google.android.exoplayer.downloadService.action.INIT";
+
+        /** Starts a download service, adding a new {@link DownloadAction} to be executed. */
+        public const string ACTION_ADD = "com.google.android.exoplayer.downloadService.action.ADD";
+
+        /** Like {@link #ACTION_INIT}, but with {@link #KEY_FOREGROUND} implicitly set to true. */
+        private const string ACTION_RESTART =
+                "com.google.android.exoplayer.downloadService.action.RESTART";
+
+        /** Starts download tasks. */
+        public const string ACTION_START_DOWNLOADS =
+                "com.google.android.exoplayer.downloadService.action.START_DOWNLOADS";
+
+        /** Stops download tasks. */
+        private const string ACTION_STOP_DOWNLOADS =
+                "com.google.android.exoplayer.downloadService.action.STOP_DOWNLOADS";
+
+        /** Key for the {@link DownloadAction} in an {@link #ACTION_ADD} intent. */
+        public const string KEY_DOWNLOAD_ACTION = "download_action";
+
+        /**
+         * Key for a bool flag in any intent to indicate whether the service was started in the
+         * foreground. If set, the service is guaranteed to call {@link #startForeground(int,
+         * Notification)}.
+         */
+        public const string KEY_FOREGROUND = "foreground";
+
+        /** Default foreground notification update interval in milliseconds. */
+        public static long DEFAULT_FOREGROUND_NOTIFICATION_UPDATE_INTERVAL = 1000;
+
+        private static string TAG = "DownloadService";
+        private static bool DEBUG = false;
+
+        // Keep the requirements helper for each DownloadService as long as there are tasks (and the
+        // process is running). This allows tasks to resume when there's no scheduler. It may also allow
+        // tasks the resume more quickly than when relying on the scheduler alone.
+        private static Dictionary<DownloadService, RequirementsHelper> requirementsHelpers = new Dictionary<DownloadService, RequirementsHelper>();
+
+        private ForegroundNotificationUpdater foregroundNotificationUpdater;
+        private string channelId;
+        private int channelName;
+
+        private DownloadManager downloadManager;
+        private DownloadManagerListener downloadManagerListener;
+        private int lastStartId;
+        private bool startedInForeground;
+
+        ///HACK Needed to instantiate the service, Java calls super() from method body, C# for some reason can't do ctor : base(args) or ctor(args) -> ctor() : this(args)
+        ///TODO Must call Initialize() from child class.
```

Comment:
Need to fix the TODO?

Reply:
Initialize needs to be called in the implementing class.